### PR TITLE
feat(api, robot-server, shared-data): Add single and multi wavelength read capabilities to the Plate Reader.

### PR DESCRIPTION
**Pull request by vegano1**
# Overview

The Byonoy Absorbance 96 Plate Reader can initialize in either `single` or `multi` measure modes, depending on the mode the device can read a single or multiple wavelengths at once. In `single` mode the device can optionally take a `reference wavelength` as a parameter, so the returned data is processed automatically on the device. This functionality was always available in the Byonoy library, this pull request exposes that API through the Opentrons hardware controller and Protocol Engine to be used in a protocol.

## Protocol API

The Protocol API for `initialize` and `read` is now

Initialize:
```def initialize(mode: ABSMeasureMode, wavelengths: List[int], reference_wavelength: Optional[int] = None) -> None```

The `initialize` method now takes a `mode: ABSMeasureMode` literal string` single` or `multi` to denote the measurement mode the absorbance reader will be configured in. The `wavelengths` parameter is a list of integers representing the wavelengths we want to measure, the plate reader needs to support these wavelengths to make measurements. The `reference_wavelength` is an optional argument used only in `single` mode and serves as a `control` wavelength the plate reader can use to perform calculations on the device when making measurements.
 
read:
`def read(self) -> Optional[Dict[int, Dict[str, float]]]`

The `read` method no longer takes any arguments but now returns a dictionary of the sample wavelengths to a dictionary of well names by absorbance values.

Closes: [PLAT-474](https://opentrons.atlassian.net/browse/PLAT-474)

## Test Plan Hands-on Testing

- [x] Make sure we can take a single reading in `single` mode
- [x] Make sure we can pass in a `reference wavelength` only in `single` mode.
- [x] Make sure we return the data for all sample wavelengths passed in for `multi` mode.
- [x] Make sure analysis fails when passing invalid values to the `initialize` AbsorbanceReaderContext API.
- [x] Make sure the following protocol analyses and runs successfully

```
from typing import cast
from opentrons import protocol_api
from opentrons.protocol_api.module_contexts import AbsorbanceReaderContext

# metadata
metadata = {
    'protocolName': 'Absorbance Reader Multi read test',
    'author': 'Platform Expansion',
}

requirements = {
    "robotType": "Flex",
    "apiLevel": "2.21",
}

# protocol run function
def run(protocol: protocol_api.ProtocolContext):
    mod = cast(AbsorbanceReaderContext, protocol.load_module("absorbanceReaderV1", "C3"))
    plate = protocol.load_labware("nest_96_wellplate_200ul_flat", "C2")
    tiprack_1000 = protocol.load_labware(load_name='opentrons_flex_96_tiprack_1000ul', location="B2")
    trash_labware = protocol.load_trash_bin("B3")
    instrument = protocol.load_instrument("flex_8channel_1000", "right", tip_racks=[tiprack_1000])
    instrument.trash_container = trash_labware

    # pick up tip and perform action
    instrument.pick_up_tip(tiprack_1000.wells_by_name()['A1'])
    instrument.aspirate(100, plate.wells_by_name()['A1'])
    instrument.dispense(100, plate.wells_by_name()['B1'])
    instrument.return_tip()
   
    # Initialize to a single wavelength with reference wavelength
    mod.initialize('single', [450], 400)

    # NOTE: CANNOT INITIALIZE WITH THE LID OPEN

    # Remove the Plate Reader lid using the Gripper.
    mod.open_lid()
    protocol.move_labware(plate, mod, use_gripper=True)
    mod.close_lid()

    # Take a reading and show the resulting absorbance values.
    result = mod.read()
    msg = f"single: {result}"
    protocol.comment(msg=msg)
    protocol.pause(msg=msg)

    # Initialize to multiple wavelengths
    protocol.pause(msg="Perform Multi Read")
    mod.open_lid()
    protocol.move_labware(plate, "C2", use_gripper=True)
    mod.close_lid()
    mod.initialize('multi', [450, 450, 450, 450, 570, 600])

    mod.open_lid()
    protocol.move_labware(plate, mod, use_gripper=True)
    mod.close_lid()

    # Take reading
    result = mod.read()
    msg = f"multi: {result}"
    protocol.comment(msg=msg)
    protocol.pause(msg=msg)

    # Place the Plate Reader lid back on using the Gripper.
    mod.open_lid()
    protocol.move_labware(plate, "C2", use_gripper=True)
    mod.close_lid()
```

## Changelog

- Add `byonoy_abs96_initialize_single_measurement` and `byonoy_abs96_initialize_multiple_measurement` to the `AsyncByonoy` class when initializing the device.
- Add `byonoy_abs96_single_measure` and `byonoy_abs96_multiple_measure` Byonoy library API to the `AsyncByonoy` class when reading absorbance.
- Add `reference_wavelenth` as an option when initializing the plate reader in `single` mode
- Expose `single` and `mult` capability to the protocol engine core and protocol engine context.
- The protocol API `read` now returns a map of wavelengths to a map of the readings to their well names.

## Review requests

- Does this make sense?
- Is the validation for the `absorbance_reader/initialization.py` command the correct location for it?
- do the doc strings need more work?

## Risk assessment
Low, unreleased

[PLAT-474]: https://opentrons.atlassian.net/browse/PLAT-474?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 1]
```diff
@@ -1003,11 +1004,28 @@ def is_lid_on(self) -> bool:
...
+        """Initialize the Absorbance Reader by taking zero reading.
+        You can initialize in `singleMeasure` or `multiMeasure` modes.
+
+        `singleMeasure` mode takes one sample wavelength and an optional
+        reference wavelength and performs a single read with the `read` method.
+
+        `multiMeasure` mode takes a list of up to 6 wavelengths and  measures the
+        absorbance on the given wavelengths when the `read` method is used.
+        Note you cannot use the `reference_wavelength` in this mode.
+        """
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/module_contexts.py]: The names of the modes strike me as a little odd within Python and the API. I would expect something like `single_measurement` and `multi_measurement`, or simply `single` and `multi` or `multiple`. Regardless, I'll keep the names as-is in my suggestion here. 

Here's a revision that mostly reformats the docstring to follow RST syntax (double backticks, whee) and include the tags for each argument of the method:

```
        """Take a zero reading on the Absorbance Plate Reader Module.
        
        :param mode: Either ``"singleMeasure"`` or ``"multiMeasure"``.

             - In single measurement mode, :py:meth:`.AbsorbanceReaderContext.read` uses 
               one sample wavelength and an optional reference wavelength.
             - In multiple measurement mode, :py:meth:`.AbsorbanceReaderContext.read` uses
               a list of up to six sample wavelengths.
        :param wavelengths: A list of wavelengths, in mm, to measure. 
             - Must contain only one item when initializing a single measurement. 
             - Must contain two to six items when initializing a multiple measurement.
        :param reference_wavelength: An optional reference wavelength, in mm. Cannot be 
             used with multiple measurements.
        """
```

This is still pretty stubby. Naturally there will be a full docs page that we can ultimately cross-reference. Before we write that, someone is going to have to ELI5 what the difference between the sample and reference wavelengths is 😆

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 2]
```diff
@@ -1003,11 +1004,28 @@ def is_lid_on(self) -> bool:
...
+        self._core.initialize(
+            mode, wavelengths, reference_wavelength=reference_wavelength
+        )
 
     @requires_version(2, 21)
-    def read(self) -> Optional[Dict[str, float]]:
+    def read(self) -> Optional[Dict[int, Dict[str, float]]]:
         """Initiate read on the Absorbance Reader. Returns a dictionary of values ordered by well name."""
+        # TODO: Update the documentation
```
Comment (ecormany) [inline api/src/opentrons/protocol_api/module_contexts.py]: Indeed! Although probably not too much to add here, since all of the user choices are front-loaded in `.initialize()`.

We should spell out any differences between the output formats for single and multi reads though.

File: api/src/opentrons/protocol_engine/commands/absorbance_reader/initialize.py [hunk 3]
```diff
@@ -54,7 +63,43 @@ async def execute(
...
+                if sample_wavelengths_len != 1:
+                    raise ValueError(
+                        f"singleMeasure requires one sample wavelength, provided {sample_wavelengths}"
+                    )
+                if (
+                    reference_wavelength is not None
+                    and reference_wavelength not in supported_wavelenths
+                ):
+                    raise ValueError(
+                        f"Reference wavelength {reference_wavelength} not supported {supported_wavelenths}"
```
Comment (ecormany) [inline api/src/opentrons/protocol_engine/commands/absorbance_reader/initialize.py]:
```suggestion
            supported_wavelengths = set(abs_reader.supported_wavelengths)
            unsupported_wavelengths = sample_wavelengths.difference(supported_wavelengths)
            if unsupported_wavelengths:
                raise ValueError(f"Unsupported wavelengths: {unsupported_wavelengths}")

            if params.measureMode == "singleMeasure":
                if sample_wavelengths_len != 1:
                    raise ValueError(
                        f"singleMeasure requires one sample wavelength, provided {sample_wavelengths}"
                    )
                if (
                    reference_wavelength is not None
                    and reference_wavelength not in supported_wavelengths
                ):
                    raise ValueError(
                        f"Reference wavelength {reference_wavelength} not supported {supported_wavelengths}"
```
😇

**Review by ecormany — COMMENTED** [review]
Some possible directions for the docstrings for these methods — provided their behavior doesn't change, of course! And I couldn't help but spot some typos.

File: api/src/opentrons/protocol_api/module_contexts.py (hunk 1 above)
Comment (vegano1) [reply api/src/opentrons/protocol_api/module_contexts.py]: @ecormany 
Thanks for the suggestions, good call out on the camel case, I've switched it to snake case.

As for the sample vs reference wavelengths

* `sample_wavelengths` are the wavelengths on which the plate reader is configured to measure absorbance. This could be 1 in `single` mode or 1-6 wavelengths in `multi` mode.
* `reference_wavelenth` is an optional baseline or control wavelength used in `single` mode, when provided the device uses that to perform calculations on the absorbance measurements. When omitted no postprocessing is performed on the device and the raw absorbance measurements are returned.

Maybe you can wordsmith that a bit into the description

File: api/src/opentrons/protocol_api/module_contexts.py (hunk 2 above)
Comment (vegano1) [reply api/src/opentrons/protocol_api/module_contexts.py]: The output is the same in both cases, will remove todo.

File: api/src/opentrons/protocol_engine/commands/absorbance_reader/read.py [hunk 4]
```diff
@@ -18,16 +18,17 @@
...
-    """Input parameters for a single absorbance reading."""
+    """Input parameters for an absorbance reading."""
 
     moduleId: str = Field(..., description="Unique ID of the Absorbance Reader.")
-    sampleWavelength: int = Field(..., description="Sample wavelength in nm.")
 
 
 class ReadAbsorbanceResult(BaseModel):
-    """Result data from running an aborbance reading, returned as a dictionary map of values by well name (eg. ("A1": 0.0, ...))."""
+    """Result data from running an aborbance reading, returned as a dictionary map of wavelenths containing a map of values by well name (eg. {450: {"A1": 0.0, ...}})."""
```
Comment (ecormany) [inline api/src/opentrons/protocol_engine/commands/absorbance_reader/read.py]:
```suggestion
    """Result data from running an aborbance reading, returned as a dictionary map of wavelengths containing a map of values by well name (eg. {450: {"A1": 0.0, ...}})."""
```
Set up a text replacement maybe 😆

**Review by ecormany — COMMENTED** [review]
New docstrings look good 👍 We'll definitely expand `read()` when we do the full documentation for the plate reader, but that's for another day.

Much other code for others to review.

File: api/src/opentrons/drivers/absorbance_reader/abstract.py [hunk 5]
```diff
@@ -32,11 +33,18 @@ async def get_available_wavelengths(self) -> List[int]:
...
+        mode: ABSMeasurementMode = ABSMeasurementMode.SINGLE,
+        reference_wavelength: Optional[int] = None,
+    ) -> None:
+        """Initialize measurement for the device in single or multi mode for the given wavelengths"""
         ...
 
     @abstractmethod
-    async def initialize_measurement(self, wavelength: int) -> None:
+    async def get_measurement(self) -> List[List[float]]:
+        """Gets an absorbance reading with the current config."""
```
Comment (CaseyBatten) [inline api/src/opentrons/drivers/absorbance_reader/abstract.py]: Given that this takes a list of wavelengths, perhaps change the doc string to reflect that this potentially returns one-or-many absorbance readings?

File: api/src/opentrons/drivers/absorbance_reader/async_byonoy.py [hunk 6]
```diff
@@ -225,23 +227,25 @@ async def get_supported_wavelengths(self) -> list[int]:
         self._supported_wavelengths = wavelengths
         return wavelengths
 
-    async def get_single_measurement(self, wavelength: int) -> List[float]:
-        """Get a single measurement based on the current configuration."""
+    async def get_measurement(self) -> List[List[float]]:
+        """Get a measurement based on the current configuration."""
```
Comment (CaseyBatten) [inline api/src/opentrons/drivers/absorbance_reader/async_byonoy.py]: Same as above, "one or many measurements" - something along those lines.

File: api/src/opentrons/drivers/absorbance_reader/async_byonoy.py [hunk 7]
```diff
@@ -225,23 +227,25 @@ async def get_supported_wavelengths(self) -> list[int]:
...
+            measure_func = self._interface.byonoy_abs96_multiple_measure
         err, measurements = await self._loop.run_in_executor(
             executor=self._executor,
             func=partial(
-                self._interface.byonoy_abs96_single_measure,
+                measure_func,
                 handle,
                 self._current_config,
             ),
         )
```
Comment (CaseyBatten) [inline api/src/opentrons/drivers/absorbance_reader/async_byonoy.py:246]: Since the `read()` call in protocol API for the plate reader is a blocking call, and is dependent upon this await down the line, we want to be pretty sure that during a read with the maximum number of wavelengths no other asynchronous processes are interrupted. From an external standpoint the error case following this will ensure a graceful "fail" if the read were to fail, killing simultaneous processes, but is there a problematic "success" case we can reach that would cause desynchronization elsewhere (read completes, but after a long time)? I think the single command single action nature of PE prevents this, but what about step-adjacent commands, like set temperature on the temperature module (which just keeps querying the poller in the background after execute)?

**Review by CaseyBatten — APPROVED** [review]
Overall looks good, thanks for tracking this one. Some comments below regarding some docstring details and a potential concern I have for the blocking behavior that our `read()` function has.

File: api/src/opentrons/drivers/absorbance_reader/async_byonoy.py (hunk 7 above)
Comment (vegano1) [reply api/src/opentrons/drivers/absorbance_reader/async_byonoy.py:246]: This is executed in a thread, so we won't be blocking any async process even in the worst-case scenario where we wait indefinitely.

File: api/src/opentrons/protocol_api/module_contexts.py [hunk 8]
```diff
@@ -1003,11 +1004,31 @@ def is_lid_on(self) -> bool:
         return self._core.is_lid_on()
 
     @requires_version(2, 21)
-    def initialize(self, wavelength: int) -> None:
-        """Initialize the Absorbance Reader by taking zero reading."""
-        self._core.initialize(wavelength)
+    def initialize(
```
Comment (ahiuchingau) [inline api/src/opentrons/protocol_api/module_contexts.py:1007]: Nit: I probably would break this out into `initialize_single_measurement` and `initialize_multiple_measurements`. This way the parameter requirements are more strict and clearer to protocol users in my opinion.

**Review by ahiuchingau — APPROVED** [review]
The implementation looks good to me!

Just left a nit on the module context `initialize` command.
